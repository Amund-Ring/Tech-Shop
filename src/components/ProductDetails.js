import React, { useEffect, useState } from "react";
import * as styles from "../styles/ProductDetails.module.css";
import { colorIsAvailable } from "../data/dataHandler";

function ProductDetails({ item, amountInCart, addItemToCart }) {
  const getStorageOptions = color => {
    //eslint-disable-next-line
    return item.options.find(option => option["color"] == color).storage;
  };

  const getPowerOptions = color => {
    //eslint-disable-next-line
    return item.options.find(option => option["color"] == color).power;
  };

  const [selectedColor, setSelectedColor] = useState(
    item.options[0]["color"].toString()
  );
  const [selectedStorage, setSelectedStorage] = useState(null);
  const [selectedPower, setSelectedPower] = useState(null);

  const [storageOptions, setStorageOptions] = useState([]);
  const [powerOptions, setPowerOptions] = useState([]);
  const [colorAvailable, setColorAvailable] = useState(
    colorIsAvailable(item.id, selectedColor)
  );

  useEffect(() => {
    const storageChoices = getStorageOptions(selectedColor);
    setStorageOptions(storageChoices);
    if (typeof storageChoices !== "undefined") {
      setSelectedStorage(storageChoices[0]);
    }
    //eslint-disable-next-line
  }, [selectedColor]);

  useEffect(() => {
    const powerChoices = getPowerOptions(selectedColor);
    setPowerOptions(powerChoices);
    if (typeof powerChoices !== "undefined") {
      setSelectedPower(powerChoices[0]);
    }
    //eslint-disable-next-line
  }, [selectedColor]);

  useEffect(() => {
    setColorAvailable(colorIsAvailable(item.id, selectedColor));
    //eslint-disable-next-line
  }, [selectedColor]);

  const handleButtonClick = () => {
    if (colorAvailable) {
      const newCartObject = {
        lineItemId: Date.now(),
        productId: item.id,
        name: item.name,
        price: item.price,
        color: selectedColor,
        storage: selectedStorage,
        power: selectedPower,
      };

      addItemToCart(newCartObject);
      setColorAvailable(colorIsAvailable(item.id, selectedColor));
    }
  };

  return (
    <div className={styles.productDetails}>
      <h2>{item.name}</h2>
      <h3>{`${item.price},-`}</h3>
      <div className={styles.imageContainer}>
        <img src={`/products/${item.id}.jpg`} alt="Alternative text" />
      </div>

      <div className={styles.details}>
        <div className={styles.detailsText}>
          <p className={styles.detailsName}>{item.name}</p>
          <p>Brand: {item.brand}</p>
          <p>Availability: {item.available ? "In stock" : "Not available"}</p>
          <p>Weight: {item.weight}kg</p>
        </div>

        <div className={styles.options}>
          <div className={styles.radioButtons}>
            <p>Colors:</p>
            <div className={styles.colorButtons}>
              {item.options.map((option, index) => {
                return (
                  <div key={index}>
                    <input
                      onClick={() => {
                        setSelectedColor(option["color"].toString());
                        // setColorAvailable(colorIsAvailable(item.id, selectedColor));
                        // console.log(selectedColor);
                      }}
                      type="radio"
                      value={option["color"]}
                      name="color"
                      defaultChecked={index === 0}
                    />{" "}
                    {option["color"]}
                  </div>
                );
              })}
            </div>
          </div>

          <div
            className={`${
              storageOptions ? styles.radioButtons : styles.hidden
            }`}
          >
            {storageOptions ? <p>Storage:</p> : ""}
            {
              <div className={styles.optionButtons}>
                {storageOptions
                  ? storageOptions.map((option, index) => {
                      return (
                        <div key={index}>
                          <input
                            onClick={() => {
                              setSelectedStorage(option.toString());
                            }}
                            type="radio"
                            value={option}
                            name="storage"
                            defaultChecked={index === 0}
                          />{" "}
                          {option} GB
                        </div>
                      );
                    })
                  : ""}
              </div>
            }
          </div>

          <div
            className={`${powerOptions ? styles.radioButtons : styles.hidden}`}
          >
            {powerOptions ? <p>Power:</p> : ""}
            {
              <div className={styles.optionButtons}>
                {powerOptions
                  ? powerOptions.map((option, index) => {
                      return (
                        <div key={index}>
                          <input
                            onClick={() => {
                              setSelectedPower(option.toString());
                              // setPowerOptionClicked(true);
                            }}
                            type="radio"
                            value={option}
                            name="power"
                            defaultChecked={index === 0}
                          />{" "}
                          {option}W
                        </div>
                      );
                    })
                  : ""}
              </div>
            }
          </div>

          {/* {item.options.map((option, index) => {
            return <div>
              <p>Color: {option["color"]}</p>
              <p>{option["storage"] ? `Storage: ${option["storage"]}` : ""}</p>
            </div>
          })} */}
        </div>
        <button
          onClick={handleButtonClick}
          className={`${
            colorAvailable ? styles.availableButton : styles.unavailableButton
          }`}
        >{`${colorAvailable ? "Add to Cart" : "Not available"}`}</button>
      </div>
    </div>
  );
}

export default ProductDetails;
