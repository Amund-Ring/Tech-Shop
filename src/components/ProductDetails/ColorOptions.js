import React from "react";
import * as styles from "../../styles/ProductDetails.module.css";

function ColorOptions({ item, setSelectedColor }) {
  return (
    <div className={styles.radioButtons}>
      <p>Colors:</p>
      <div className={styles.colorButtons}>
        {item.options.map((option, index) => {
          return (
            <div key={index}>
              <input
                onClick={() => {
                  setSelectedColor(option["color"].toString());
                }}
                type="radio"
                value={option["color"]}
                name="color"
                defaultChecked={index === 0}
              />{" "}
              {option["color"]}
            </div>
          );
        })}
      </div>
    </div>
  );
}

export default ColorOptions;
