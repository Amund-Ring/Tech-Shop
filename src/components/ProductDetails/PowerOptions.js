import React from "react";
import * as styles from "../../styles/ProductDetails.module.css";

function PowerOptions({ powerOptions, setSelectedPower }) {
  return (
    <div className={`${powerOptions ? styles.radioButtons : styles.hidden}`}>
      {powerOptions ? <p>Power:</p> : ""}
      {
        <div className={styles.optionButtons}>
          {powerOptions
            ? powerOptions.map((option, index) => {
                return (
                  <div key={index}>
                    <input
                      onClick={() => {
                        setSelectedPower(option.toString());
                      }}
                      type="radio"
                      value={option}
                      name="power"
                      defaultChecked={index === 0}
                    />{" "}
                    {option}W
                  </div>
                );
              })
            : ""}
        </div>
      }
    </div>
  );
}

export default PowerOptions;
