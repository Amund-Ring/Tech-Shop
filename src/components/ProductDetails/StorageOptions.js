import React from "react";
import * as styles from "../../styles/ProductDetails.module.css";

function StorageOptions({ item, storageOptions, setSelectedStorage }) {
  return (
    <div className={`${storageOptions ? styles.radioButtons : styles.hidden}`}>
      {storageOptions ? <p>Storage:</p> : ""}
      {
        <div className={styles.optionButtons}>
          {storageOptions
            ? storageOptions.map((option, index) => {
                return (
                  <div key={index}>
                    <input
                      onClick={() => {
                        setSelectedStorage(option.toString());
                      }}
                      type="radio"
                      value={option}
                      name="storage"
                      defaultChecked={index === 0}
                    />{" "}
                    {option} GB
                  </div>
                );
              })
            : ""}
        </div>
      }
    </div>
  );
}

export default StorageOptions;
