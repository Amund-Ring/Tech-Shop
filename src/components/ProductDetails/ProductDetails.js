import React, { useEffect, useState } from "react";
import * as styles from "../../styles/ProductDetails.module.css";
import { colorIsAvailable } from "../../data/dataHandler";
import ColorOptions from "./ColorOptions";
import PowerOptions from "./PowerOptions";
import StorageOptions from "./StorageOptions";

function ProductDetails({ item, amountInCart, addItemToCart }) {
  const getStorageOptions = color => {
    //eslint-disable-next-line
    return item.options.find(option => option["color"] == color).storage;
  };

  const getPowerOptions = color => {
    //eslint-disable-next-line
    return item.options.find(option => option["color"] == color).power;
  };

  const [selectedColor, setSelectedColor] = useState(
    item.options[0]["color"].toString()
  );
  const [selectedStorage, setSelectedStorage] = useState(null);
  const [selectedPower, setSelectedPower] = useState(null);

  const [storageOptions, setStorageOptions] = useState([]);
  const [powerOptions, setPowerOptions] = useState([]);
  const [colorAvailable, setColorAvailable] = useState(
    colorIsAvailable(item.id, selectedColor)
  );

  useEffect(() => {
    const storageChoices = getStorageOptions(selectedColor);
    setStorageOptions(storageChoices);
    if (typeof storageChoices !== "undefined") {
      setSelectedStorage(storageChoices[0]);
    }
    //eslint-disable-next-line
  }, [selectedColor]);

  useEffect(() => {
    const powerChoices = getPowerOptions(selectedColor);
    setPowerOptions(powerChoices);
    if (typeof powerChoices !== "undefined") {
      setSelectedPower(powerChoices[0]);
    }
    //eslint-disable-next-line
  }, [selectedColor]);

  useEffect(() => {
    setColorAvailable(colorIsAvailable(item.id, selectedColor));
    //eslint-disable-next-line
  }, [selectedColor]);

  const handleButtonClick = () => {
    if (colorAvailable) {
      const newCartObject = {
        lineItemId: Date.now(),
        productId: item.id,
        name: item.name,
        price: item.price,
        color: selectedColor,
        storage: selectedStorage,
        power: selectedPower,
      };

      addItemToCart(newCartObject);
      setColorAvailable(colorIsAvailable(item.id, selectedColor));
    }
  };

  return (
    <div className={styles.productDetails}>
      <h2>{item.name}</h2>
      <h3>{`${item.price},-`}</h3>
      <div className={styles.imageContainer}>
        <img src={`/products/${item.id}.jpg`} alt={item.name} />
      </div>

      <div className={styles.details}>
        <div className={styles.detailsText}>
          <p className={styles.detailsName}>{item.name}</p>
          <p>Brand: {item.brand}</p>
          <p>Availability: {item.available ? "In stock" : "Not available"}</p>
          <p>Weight: {item.weight}kg</p>
        </div>

        <div className={styles.options}>
          <ColorOptions item={item} setSelectedColor={setSelectedColor} />

          <PowerOptions
            item={item}
            powerOptions={powerOptions}
            setSelectedPower={setSelectedPower}
          />

          <StorageOptions
            item={item}
            storageOptions={storageOptions}
            setSelectedStorage={setSelectedStorage}
          />
        </div>

        <button
          onClick={handleButtonClick}
          className={`${
            colorAvailable ? styles.availableButton : styles.unavailableButton
          }`}
        >{`${colorAvailable ? "Add to Cart" : "Not available"}`}</button>
      </div>
    </div>
  );
}

export default ProductDetails;
