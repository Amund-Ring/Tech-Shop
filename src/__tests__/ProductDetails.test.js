import React from "react";
import renderer from "react-test-renderer";
import ProductDetails from "../components/ProductDetails/ProductDetails";

describe("ProductDetails", () => {
  it("renders correctly", () => {
    const item = {
      id: 1,
      name: "Philips hue bulb",
      brand: "Philips",
      price: "500",
      available: true,
      weight: 0.2,
      options: [
        {
          color: "white",
          power: [6.5, 9.5],
          quantity: 3,
        },
        {
          color: "red",
          power: [6.5, 9.5],
          quantity: 7,
        },
      ],
    };

    const tree = renderer
      .create(<ProductDetails item={item} amountInCart={3} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
