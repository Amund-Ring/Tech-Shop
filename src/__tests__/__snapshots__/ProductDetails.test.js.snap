// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetails renders correctly 1`] = `
<div>
  <h2>
    Philips hue bulb
  </h2>
  <h3>
    500,-
  </h3>
  <div>
    <img
      alt="Philips hue bulb"
      src="/products/1.jpg"
    />
  </div>
  <div>
    <div>
      <p>
        Philips hue bulb
      </p>
      <p>
        Brand: 
        Philips
      </p>
      <p>
        Availability: 
        In stock
      </p>
      <p>
        Weight: 
        0.2
        kg
      </p>
    </div>
    <div>
      <div>
        <p>
          Colors:
        </p>
        <div>
          <div>
            <input
              defaultChecked={true}
              name="color"
              onClick={[Function]}
              type="radio"
              value="white"
            />
             
            white
          </div>
          <div>
            <input
              defaultChecked={false}
              name="color"
              onClick={[Function]}
              type="radio"
              value="red"
            />
             
            red
          </div>
        </div>
      </div>
      <div
        className="undefined"
      >
        <p>
          Power:
        </p>
        <div />
      </div>
      <div
        className="undefined"
      >
        <p>
          Storage:
        </p>
        <div />
      </div>
    </div>
    <button
      className="undefined"
      onClick={[Function]}
    >
      Add to Cart
    </button>
  </div>
</div>
`;
